designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+This is equivalent to the [operator framework](https://operatorframework.io/operator-capabilities) level 2 .
+
+The rest of this document is a high level specification about changes to the EKSA controller and CLI necessary to allow users to create/upgrade/delete EKS Anywhere clusters using Kubectl or GitOps.
+
+## Solution Details
+
+### Architectural Diagram
+
+![Cluster lifecycle](images/cluster-lifecycle.png)
+### Cluster Reconciler
+
+The main logic for this feature will be implemented in the cluster reconciler. This will take care of creating/patching and deleting all resources needed for a cluster to be fully operational, from CAPI resources in the management cluster to installing the CNI in the workload cluster.
+
+This flow will be very similar to what we do today in the CLI: a mix between the create and upgrade flows. However, the "wait for condition" pattern that we extensively use in the CLI is not suitable for an event based system like a Kubernetes controller. We shouldn't block the execution of the control loop. This could be done instead, by watching resources from the reconciler (eg. CAPI `Cluster` and `KubeadmControlPlane`) and use an event handler to enqueue cluster reconcile requests in response to those events.
+
+![Cluster reconcile](images/cluster_reconcile.png)
+
+Note: this diagram is missing things like applying tolerations to a vSphere controller, taints for worker nodes, force updating the `ClusterResourceSet` resources, IAM auth, etc. Those will be added as we start the implementation, when we decide where they should go in the flow. 
+
+
+In order to check if something needs to change, reconcilers are supposed to compare the desired state (in our case, a combination of multiple EKSA objects, like `Cluster`, `<Provider>MachineConfig`, etc.) with the current state (in our case a combination of CAPI objects, CNI and some extra Kubernetes objects, depending on the provider).
+
+Following that, on every reconcile loop we would compare (object by object) if the objects coming from our CAPI "template" are different than the ones already present in the clusters (some of them will be in the management cluster and others in the workload cluster).
+
+This could be an expensive operation, so we could potentially cache it in the status of our cluster object. We would need a combination of all the observed generation's for all the resources that represent our desired state (cluster itself, provider datacenter, machineconfigs...) and the "phase" (for example, "CP reconciled"). A drawback of this approach is that, once the cluster had passed a certain phase, if any of the "result" objects changes (for example by manual user interaction), we would no "reconcile them back" to the desired state. Such reconciliation would only happen upon change in our EKSA objects. We won't implement this right away and will only reconsider if we find performance issues.
```

Comment:
```suggestion
This could be an expensive operation, so we could potentially cache it in the status of our cluster object. We would need a combination of all the observed generation's for all the resources that represent our desired state (cluster itself, provider datacenter, machineconfigs...) and the "phase" (for example, "CP reconciled"). A drawback of this approach is that, once the cluster had passed a certain phase, if any of the "result" objects changes (for example by manual user interaction), we would not be able to "reconcile them back" to the desired state. Such reconciliation would only happen upon change in our EKSA objects. We won't implement this right away and will only reconsider if we find performance issues.
```

Reply:
When we implement this we can figure out how much performance we lose. Otherwise the other option would be is to write all the hooks for reconciling with the capi objects, which would be so hard to maintain but avoids just looking at the entire template. I still support using the template to compare the objects and let capi reconcile, but we can think about breaking the template further in smaller pieces to be able to not have to reconcile all the objects at the same time. I think you touch on this on the below paragraph too.

Reply:
Yeah, I was thinking we can probably add some logging with the time it takes to run each reconcile" step.

To be honest, we don't think we'll need to do anything about performance, at least during these first iterations. From a user's expectations point of view, cluster reconciliation is no something that needs to happen immediately. Obviously, the faster the better, but taking a few tens of seconds extra is not a big deal.